File: built-in plugins dsp/mb-creative-spectral-gate.cpp
/**
 * MB Spectral Gate
 * Category : effect
 * Type     : gate
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : FFT-based spectral gating for isolating specific frequency regions
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_CREATIVE_SPECTRAL_GATE_H
#define MB_CREATIVE_SPECTRAL_GATE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCreativeSpectralGate : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-creative-spectral-gate";
    static constexpr const char* PLUGIN_NAME    = "MB Spectral Gate";
    static constexpr const char* PLUGIN_TYPE    = "gate";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -40f;  // range [-80, 0]
    float fftSize = 2048f;  // range [256, 8192]
    float smoothing = 0.5f;  // range [0, 1]
    float mix = 1f;  // range [0, 1]
    };

    MbCreativeSpectralGate() = default;
    ~MbCreativeSpectralGate() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -80f, 0f);
        params.fftSize = std::clamp(params.fftSize, 256f, 8192f);
        params.smoothing = std::clamp(params.smoothing, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Spectral Gate
        return input;
    }
};

#endif // MB_CREATIVE_SPECTRAL_GATE_H

File: built-in plugins dsp/mb-creative-spectral-gate.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbCreativeSpectralGatePlugin: PluginDefinition = { id: 'mb-creative-spectral-gate', slug: 'mb-creative-spectral-gate', name: 'MB Spectral Gate', category: 'effect', type: 'gate' as any, version: '1.0.0', description: 'FFT-based spectral gating for isolating specific frequency regions', author: 'Max Booster', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: -40, minValue: -80, maxValue: 0, automatable: true }, { id: 'fftSize', name: 'FFT Size', type: 'float', defaultValue: 2048, minValue: 256, maxValue: 8192, automatable: false }, { id: 'smoothing', name: 'Smoothing', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { threshold: -40, fftSize: 2048, smoothing: 0.5, mix: 1 } };

export default MbCreativeSpectralGatePlugin;
